[torchgen] Refactor static_dispatch to take in source signature

larryliu0820 opened the pull request:
Summary: Context: currently `static_dispatch` assumes that given a native function `f`, we always want to map from its `DispatchSignature` to its `CppSignature`. This assumption may not hold true for some use cases, where the source bindings may not come from its `DispatchSignature`. Here I'm changing the argument `sig: DispatcherSignature` to be `sig: Union[CppSignature, DispatcherSignature]`, also removes unused `f` 

Test Plan: Rely on added unit test.

Differential Revision: D39192969

ezyang commented on torchgen/gen.py:
```diff
@@ -434,7 +437,19 @@ def generate_static_dispatch_fallback_call(
 def static_dispatch(
```
The function overall here looks a bit suspect. You are assuming ambient bindings from the sig context at the call site; why isn't that sig passed as an argument here?

iseeyuan commented on tools/test/test_codegen.py:
```diff
@@ -9,10 +9,14 @@
 import yaml
 
 from tools.autograd import gen_autograd_functions, load_derivatives
+from torchgen.api import cpp
```
Remove?

iseeyuan reviewed (approved):
LGTM. Thanks for unblocking Executorch in ATen mode!

ezyang commented on torchgen/gen.py:
```diff
@@ -359,11 +360,13 @@ def static_dispatch_extra_headers(backends: List[BackendIndex]) -> List[str]:
+# tools.codegen.api.translate() yet as its application is limited to static dispatch.
+def translate_args(
     f: NativeFunction,
+    cpp_sig: Optional[CppSignature],
```
it is better make this not Optional; it is an antipattern to repeat construction of sig on the inside, that means there is not one canonical source of truth

ezyang commented on torchgen/gen.py:
```diff
@@ -385,27 +388,31 @@ def add_spl_memory_format_binding(input_bindings: List[Binding]) -> List[Binding
+        list(cpp_sig.arguments())
+        if cpp_sig
+        else DispatcherSignature.from_schema(f.func).arguments()
+    )
```
and then you wouldn't have to do this

ezyang commented on torchgen/gen.py:
```diff
@@ -385,27 +388,31 @@ def add_spl_memory_format_binding(input_bindings: List[Binding]) -> List[Binding
+    for arg in goals:
         if arg.nctype.name == SpecialArgName.possibly_redundant_memory_format:
-            disp_bindings = add_spl_memory_format_binding(disp_sig.arguments())
+            src_bindings = add_spl_memory_format_binding(src_bindings)
```
Can you explain to me why this is needed? Translate knows how to handle redundant memory format correctly, so it is surprising to me that you need to paper over it more here. And even if it didn't, I would expect you to teach translate how to handle it adequately.

larryliu0820 replied:
Yeah good point I'll see what's going on with memory_format. I can follow up in another PR.

larryliu0820 committed:
[torchgen] Refactor static_dispatch to take in source bindings (#84384)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84384

Context: currently `static_dispatch` assumes that given a native function `f`, we always want to map from its `DispatchSignature` to its `CppSignature`. This assumption may not hold true for some use cases, where the source bindings may not come from its `DispatchSignature`. Here I'm adding another argument to `static_dispatch` so that it takes a list of bindings as source, then call `translate()` to map to its `CppSignature`.

Test Plan:
Rely on added unit test.

**Static Docs Preview: executorch**
|[Full Site](https://our.intern.facebook.com/intern/staticdocs/eph/D39192969/V9/executorch/)|

|**Modified Pages**|

Reviewed By: iseeyuan

Differential Revision: D39192969

fbshipit-source-id: afc7a8c7427de66d8d43fcb859e5cc359651e4aa

ezyang commented on tools/test/test_codegen.py:
```diff
@@ -409,6 +412,67 @@ def test_functional_variant_autogen_out_variant_two_returns(self) -> None:
 
 
+# Test for static_dispatch
+class TestStaticDispatchGeneratrion(unittest.TestCase):
```
Generation

larryliu0820 commented:
@pytorchbot merge
